### PR TITLE
Memory efficient DBSCAN

### PR DESCRIPTION
**Pull request by OmriRoames**
* Uses adjacency list rather than adjacency matrix (see #58 )
* Suports PointCloud container
* keeps track of core and boundary points

Updated pull request instead of #61

### CODE REVIEW

File: src/dbscan.jl [hunk 1]
```diff
@@ -102,4 +109,113 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
     return cnt
 end
 
+"""
+    dbscan(points, radius [; leafsize = 20, min_neighbors = 1, min_cluster_size = 1]) -> clusters
```
Comment (ararslan) [inline src/dbscan.jl]: No need for brackets around keyword arguments

File: src/dbscan.jl [hunk 2]
```diff
@@ -15,6 +15,13 @@ type DbscanResult <: ClusteringResult
 end
 
 
+immutable DbscanCluster <: ClusteringResult
```
Comment (ararslan) [inline src/dbscan.jl]: IMO since DBSCAN is an acronym, it should be in all caps, i.e. `DBSCANCluster`

File: src/dbscan.jl [hunk 3]
```diff
@@ -102,4 +109,113 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
...
+            _update_exploration_list!(adj_list, to_explore, visited)
+        end
+        cluster_size = sum(cluster_selection)
+        min_cluster_size <= cluster_size && _accept_cluster!(clusters, core_selection, cluster_selection, cluster_size)
+    end
+    return clusters
+end
+
+"""
+`_update_exploration_list!(adj_list, exploration_list, visited)`
```
Comment (ararslan) [inline src/dbscan.jl]: Should indent 4 spaces rather than use backticks in docstrings

File: src/dbscan.jl [hunk 4]
```diff
@@ -102,4 +109,113 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
     return cnt
 end
 
+"""
+    dbscan(points, radius [; leafsize = 20, min_neighbors = 1, min_cluster_size = 1]) -> clusters
+
+Run DBSCAN spatial clustering for the given `points`.
```
Comment (ararslan) [inline src/dbscan.jl]: > Cluster `points` using the DBSCAN (density-based spatial clustering of applications with noise) algorithm.

?

File: REQUIRE [hunk 5]
```diff
@@ -1,4 +1,6 @@
 julia 0.4
+Compat 0.8.5
 Distances 0.3.1
+FixedSizeArrays
 StatsBase 0.9.0
-Compat 0.8.5
+PointClouds
```
Comment (ararslan) [inline REQUIRE]: So the dependency on PointClouds is new? I'm not familiar with it. Is it actively maintained? If it's only being used for one function (which also has another method) then it's probably not worth having another dependency.

Also should specify minimum versions for all package requirements.

Comment (ararslan) [inline REQUIRE]: PointClouds isn't registered in METADATA, which is causing the failures.

File: src/dbscan.jl [hunk 6]
```diff
@@ -102,4 +109,113 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
...
+### Output
+* `Vector{DbscanCluster}`: an array of clusters with the id, size core indices and boundary indices
+
+### Example:
+points = randn(3, 10000)
+clusters = dbscan(points, 0.05, min_neighbors = 3, min_cluster_size = 20) # clusters with less than 20 points will be discarded
+"""
+function dbscan{T <: AbstractFloat, N}(points::Array{T, N}, radius::Real; leafsize::Int = 20, kwargs ...)
+    kdtree = KDTree(points; leafsize=leafsize)
+    kdtree::KDTree{Float64, Distances.Euclidean} # TODO: julia-0.4 fails type inference here
```
Comment (KristofferC) [inline src/dbscan.jl]: Hmm I should fix this in NearestNeighbors...

File: REQUIRE (hunk 5 above)
Comment (KristofferC) [inline REQUIRE]: Need to add NearestNeighbors here as well.

File: src/dbscan.jl [hunk 7]
```diff
@@ -102,4 +109,113 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
...
+    dbscan(points, radius [; leafsize = 20, min_neighbors = 1, min_cluster_size = 1]) -> clusters
+
+Run DBSCAN spatial clustering for the given `points`.
+
+### Arguments
+* `points`: matrix of points or a `PointCloud`
+* `radius::Real`: query radius
+
+### Keyword Arguments
+* `leafsize::Int`: number of leafs in the `KDTree`
```
Comment (KristofferC) [inline src/dbscan.jl]: It is more like the number if points binned in each leaf node.

File: src/dbscan.jl [hunk 8]
```diff
@@ -102,4 +109,113 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
...
+* `min_neighbors::Int`: minimum number of neighbors to be a core point
+* `min_cluster_size::Int`: minimum number of points to be a valid cluster
+
+### Output
+* `Vector{DbscanCluster}`: an array of clusters with the id, size core indices and boundary indices
+
+### Example:
+points = randn(3, 10000)
+clusters = dbscan(points, 0.05, min_neighbors = 3, min_cluster_size = 20) # clusters with less than 20 points will be discarded
```
Comment (ararslan) [inline src/dbscan.jl]: Should wrap the example code in triple backticks as in GitHub markdown

File: REQUIRE (hunk 5 above)
Comment (OmriRoames) [inline REQUIRE]: Yes, PointClouds.jl is actively maintained by FugroRoames. It's a container for large point clouds, and there is no real reason to have it here. My bad.

**Comment by OmriRoames** [comment]
Thanks a lot for the feedback. I've removed the PointClouds dependency

File: src/dbscan.jl [hunk 9]
```diff
@@ -203,19 +199,19 @@ function _update_exploration_list!(adj_list::Array{Int}, exploration_list::Vecto
 end
 
 """
-`_accept_cluster!(clusters, core_selection, cluster_selection)`
+    `accept_cluster!(clusters, core_selection, cluster_selection)`
```
Comment (ararslan) [inline src/dbscan.jl]: When it's indented, it shouldn't still be surrounded in backticks

**Comment by OmriRoames** [comment]
Ok, fixed

**Comment by ararslan** [comment]
Looks good to me.

File: src/dbscan.jl [hunk 10]
```diff
@@ -102,4 +109,109 @@ function _dbs_expand_cluster!{T<:Real}(D::DenseMatrix{T},           # distance m
...
+
+### Example:
+``` julia
+points = randn(3, 10000)
+clusters = dbscan(points, 0.05, min_neighbors = 3, min_cluster_size = 20) # clusters with less than 20 points will be discarded
+```
+"""
+function dbscan{T <: AbstractFloat, N}(points::Array{T, N}, radius::Real; leafsize::Int = 20, kwargs ...)
+    kdtree = KDTree(points; leafsize=leafsize)
+    kdtree::KDTree{Float64, Distances.Euclidean} # TODO: julia-0.4 fails type inference here
```
Comment (KristofferC) [inline src/dbscan.jl]: Should it be a `T` here instead of `Float64`?

Comment (OmriRoames) [inline src/dbscan.jl]: Good point.

**Comment by JoshChristie** [comment]
Should the prior dbscan method be removed/deprecated?
Could you update the docs page with the new function?

**Comment by OmriRoames** [comment]
@ararslan how can I update the documentation?

**Comment by ararslan** [comment]
You can edit doc/source/dbscan.rst

File: src/Clustering.jl [hunk 11]
```diff
@@ -33,7 +34,7 @@ module Clustering
     AffinityPropResult, affinityprop,
 
     # dbscan
-    DbscanResult, dbscan,
+    DbscanResult, DBSCANCluster, dbscan,
```
Comment (ararslan) [inline src/Clustering.jl]: Just noticed: `DBSCANResult` should be capitalized to match `DBSCANCluster`.

**Comment by ararslan** [comment]
@JoshChristie If I'm not mistaken (perhaps Omri can correct me if I am), this change will have no user-facing effects aside from renaming changing the capitalization of types.

**Comment by OmriRoames** [comment]
@ararslan I like DBSCANResult better as well but I am worried that changing it will break peoples code. What do you think of @JoshChristie suggestion to deprecate the old DBSCAN?

**Comment by ararslan** [comment]
Oh my point about non-user facing was that there would be nothing to deprecate. Or have you changed the actual API of the `dbscan` function? For whatever reason it isn't showing any removed code in the diff when I look at it, but there's current a `dbscan` function on master.

**Comment by OmriRoames** [comment]
I haven't removed the existing one. The API for the new function is different: the existing one takes an adjacency matrix (which is what cause the memory to blow up) while the new one simply takes the points themselves (no adjacency matrix involved). I think the existing one should go, but if we remove it it will break peoples code. I would go with deprecation.

**Comment by ararslan** [comment]
Ah okay, I see now. I apologize for my initial confusion.

Deprecating doesn't seem necessary to me; if people want to use the method with adjacency matrices then they can, or they can use your method. I'd be interested to hear what others think about that though. @nalimilan?

**Comment by JoshChristie** [comment]
If both methods are kept it would good to recommend the new method other the old method due to the performance increase for large data.

A quick performance test:
```julia
using Clustering
# Create 3 separate clusters (30,000 points)
a = randn(3,10000)
b = randn(3,10000) .+ 5.0
c = randn(3,10000) .+ 10.0
points = [a b c]

# old method (timings below recorded after a couple of reruns)
using Distances
@time dist_matrix = pairwise(Euclidean(), points, points) 
# 30.865760 seconds (180.01 k allocations: 6.712 GB, 5.62% gc time)
@time dbscan(dist_matrix, 1.0, 1) 
# 2.843919 seconds (380.61 k allocations: 607.347 MB, 2.72% gc time)

# new method
@time dbscan(points, 1.0)
# 1.713170 seconds (501.31 k allocations: 565.157 MB, 1.83% gc time)
````

**Comment by OmriRoames** [comment]
OK, I've changed `DBSCANCluster` back to `DbscanCluster` for consistency, and wrote some documentation.

**Comment by OmriRoames** [comment]
@ararslan is there anything else that needs some attention?

**Comment by ararslan** [comment]
Don't think so. This looks great, thanks for contributing!

**Pull request merged**